assets/js/leaves.js

```diff
@@ -0,0 +1,566 @@
+// logic to generate a path for drawing leaves,
+// and for managing leaf details and tag behavior in the timetree
+
+import { select, selectAll } from "d3-selection";
+import { line, curveNatural } from "d3-shape";
+
+// combine into d3 object for convenience
+const d3 = {
+  line,
+  curveNatural,
+  select,
+  selectAll,
+};
+
+// configuration for leaf sizes
+// sizes are scaled from 40px width design:
+//   min height and width: 40px; max height 100px;
+//   max mid shift 34 (34px at most, 17 either way); max tip shift 12
+
+let leafSize = {
+  minHeight: 30,
+  maxHeight: 75,
+  width: 30,
+  maxMidShift: 25,
+  maxTipShift: 9,
+};
+
+// get a random number between a min and max
+function randomNumBetween(max, min = 0) {
+  // scale random value between 0 and 1 to desired scale, start at min value
+  return Math.random() * (max - min) + min;
+}
+
+function plusOrMinus(x) {
+  return randomNumBetween(x, -x);
+}
+
+function cointoss() {
+  // randomly return true or false;
+  // based on https://stackoverflow.com/a/60322877/9706217
+  return Math.random() < 0.5;
+}
+
+const TagSelectEvent = new Event("tag-select");
+const TagDeselectEvent = new Event("tag-deselect");
+
+class Leaf {
+  // constant for selection classname
+  static selectedClass = "select";
+  static highlightClass = "highlight";
+
+  constructor(panel, ignore_ids, tags) {
+    // store a reference to the panel object
+    this.panel = panel;
+    this.container = document.querySelector("aside");
+    this.bindHandlers();
+    // list of ids to ignore when loading leaf details
+    // (known non-leaf elements, including branch start targets)
+    this.ignore_ids = ignore_ids || [];
+    this.tags = tags || [];
+  }
+
+  static isTag(element) {
+    // check if an element is a tag
+    return (
+      element.tagName == "A" && element.parentElement.classList.contains("tags")
+    );
+  }
+
+  bindHandlers() {
+    // bind a delegated click handler to override tag link behavior;
+    // delegated so it applies to tags in leaf details loaded after bound
+    this.container.addEventListener("click", (event) => {
+      let element = event.target;
+      // if click target is a link in the tags section, select leaves for that tag
+      if (Leaf.isTag(element)) {
+        event.preventDefault();
+        event.stopPropagation();
+        this.currentTag = element.dataset.tag;
+        element.classList.add(Leaf.selectedClass);
+        this.container.dispatchEvent(TagSelectEvent);
+      }
+    });
+
+    // bind handler to current tag x button to deactivate tag
+    this.activeTagClose = document.querySelector("#current-tag .close");
+    if (this.activeTagClose) {
+      this.activeTagClose.addEventListener("click", (event) => {
```

Comment:
This may be an event worth turning off for tags that are no longer active. If a user clicks on the same tag after removing it, a new eventListener will be created and the previous one will still be in memory!
I think it should be possible to do so here, but you won't be able to use an anonymous function. Here is a guess (not tested):

```js
if (this.activeTagClose) {
  const handleTagDeselect = (event) => {
    this.currentTag = null;
    this.container.dispatchEvent(TagDeselectEvent);
    this.activeTagClose.removeEventListener("click", handleTagDeselect)
  };
  this.activeTagClose.addEventListener("click", handleTagDeselect)
}
```

Reply:
I had to look back at the template code - the active tag close button is always there, it's just not visible when there is no active tag. So this shouldn't ever create any new event listeners, and it's only bound once when the page is loaded.  But I did notice I have a delegated click handler just above that and wonder if it would make sense to consolidate them....